### PR TITLE
Replace ref.invalidate with subscribeStream

### PR DESCRIPTION
**Pull request by bitfriend**
Will fix [this issue](https://github.com/acterglobal/a3/issues/1572)

### CODE REVIEW

**Comment by bitfriend** [comment]
Found `ref.invalidate()` is called unnecessarily for `spaceEventsProvider`, even though it is not space.
Removed it.

**Comment by bitfriend** [comment]
Now user can monitor that another device is logged in by https://github.com/acterglobal/a3/pull/1598/commits/ed17a5291e709cdc4db5586db06ea08df74604b9

File: native/core/src/executor.rs [hunk 1]
```diff
@@ -101,7 +101,6 @@ impl Executor {
             }
             Ok(keys) => {
                 trace!(?event_id, "handling done");
-                info!("******************** executor handled: {:?}", keys.clone());
```
Comment (gnunicorn) [inline native/core/src/executor.rs]: why are you removing all of these? They are useful for debugging, which is the reason they are thre? Arguably the "***" can be removed but why take them out entirely?

File: native/acter/src/api/verification.rs [hunk 2]
```diff
@@ -1014,7 +1015,10 @@ impl SessionManager {
                             session: info.session.clone(),
                         });
                         let auth_data = AuthData::Password(pass_data);
-                        client.delete_devices(&devices, Some(auth_data)).await?;
```
Comment (gnunicorn) [inline native/acter/src/api/verification.rs:1017]: these changes seem unrelated to the description of this PR. Or I don't understand their relationship. please explain.

File: native/acter/src/api/device.rs [hunk 3]
```diff
@@ -90,21 +95,19 @@ impl DeviceController {
...
                 }
                 if let Some(user_devices) = device_updates.changed.get(my_id) {
-                    for device in user_devices.values() {
-                        let dev_id = device.device_id().to_owned();
+                    for (dev_id, dev) in user_devices {
                         info!("device-changed device id: {}", dev_id);
-                        let evt = DeviceChangedEvent::new(&client, dev_id);
-                        if let Err(e) = changed_event_tx.try_send(evt) {
+                        let evt = DeviceChangedEvent::new(&client, dev_id.to_owned());
+                        if let Err(e) = changed_tx.try_send(evt) {
```
Comment (gnunicorn) [inline native/acter/src/api/device.rs:110]: reading the subscriber on the flutter side, there are two clear improvements here:
1. we don't care about the details of the event on the flutter side, all that happens is that we refresh the entire list
2. it's not needed to be done _for each_ device that changed, but once for every change reported. That's totally enough.

This can and should be vastly simplified.

File: native/acter/src/api/device.rs [hunk 4]
```diff
@@ -66,15 +66,20 @@ impl DeviceChangedEvent {
 
 #[derive(Clone, Debug)]
 pub(crate) struct DeviceController {
+    new_event_tx: Sender<DeviceNewEvent>, // keep it resident in memory
```
Comment (gnunicorn) [inline native/acter/src/api/device.rs:69]: I don't think we prompt anymore for new-device detections (as they were very buggy and reported even the device I was just logging myself in-with). Not sure this features is still needed or shouldn't be completely removed then.

**Review by gnunicorn — APPROVED** [review]
There are some things I'd like to see changed.

And I'd like this to be merged not before Thursday/the next full release, as there are many `invalidate` removed that were there for UX convinience and I am not sure they aren't all warranted. Are these all tested for? Would be annoying to have run into regressions because of that and then have to add them again one-by-one :( ...

File: app/lib/common/widgets/chat/convo_card.dart [hunk 5]
```diff
@@ -134,12 +134,6 @@ class _ConvoCardState extends ConsumerState<ConvoCard> {
     await room.unmute();
     if (!mounted) return;
     EasyLoading.showToast(L10n.of(context).notificationsUnmuted);
-    await Future.delayed(const Duration(seconds: 1), () {
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply app/lib/common/widgets/chat/convo_card.dart:137]: Frankly I deleted this code, because I confirmed `maybeRoomProvider` is auto-updated without `ref.invalidate()`.
Of course, it was confirmed at another place.

File: app/lib/features/auth/providers/notifiers/auth_notifier.dart [hunk 6]
```diff
@@ -17,7 +16,6 @@ class AuthStateNotifier extends StateNotifier<bool> {
 
   Future<void> nuke(BuildContext context) async {
     await ActerSdk.nuke();
-    ref.invalidate(spacesProvider);
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply app/lib/features/auth/providers/notifiers/auth_notifier.dart:20]: You are correct.
Actually I thought forced update of spaces provider would be meaningless because `nuke()` is called when user was logged out and he is going to delete old caches.

File: app/lib/features/events/providers/event_providers.dart [hunk 7]
```diff
@@ -218,8 +221,29 @@ Future<List<ffi.CalendarEvent>> _sortEventListDscTime(
...
+  () => AsyncRsvpStatusNotifier(),
+);
+
+class AsyncRsvpStatusNotifier
+    extends AutoDisposeFamilyAsyncNotifier<ffi.OptionRsvpStatus, String> {
+  late Stream<bool> _listener;
+
+  Future<ffi.OptionRsvpStatus> _getMyResponse() async {
+    final client = ref.read(alwaysClientProvider);
+    final calEvent = await client.waitForCalendarEvent(arg, null);
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply app/lib/features/events/providers/event_providers.dart:235]: This code was already tested.
Please review this video.

https://github.com/user-attachments/assets/7f09eeed-4f76-459c-9e1b-3262329518cd

[the remark this answers is not in the document]
Comment (gnunicorn) [reply app/lib/features/events/providers/event_providers.dart:235]: not saying it doesn't work, but that the structure of the code should be improved.

File: app/lib/features/room/widgets/notifications_settings_tile.dart [hunk 8]
```diff
@@ -161,12 +161,6 @@ class _NotificationSettingsTile extends ConsumerWidget {
         return;
       }
       EasyLoading.showToast(L10n.of(context).notificationStatusSubmitted);
-      await Future.delayed(const Duration(seconds: 1), () {
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply app/lib/features/room/widgets/notifications_settings_tile.dart]: Fixed by https://github.com/acterglobal/a3/pull/1598/commits/96e14461f975f573fe0c69c947ba738bd9c70c10

File: app/lib/common/widgets/chat/convo_card.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (bitfriend) [reply app/lib/common/widgets/chat/convo_card.dart]: Fixed by https://github.com/acterglobal/a3/pull/1598/commits/96e14461f975f573fe0c69c947ba738bd9c70c10

File: app/lib/features/events/providers/event_providers.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (bitfriend) [reply app/lib/features/events/providers/event_providers.dart:235]: @kumarpalsinh25 is improving this logic in [his PR](https://github.com/acterglobal/a3/pull/1990).
I discussed with @gnunicorn and decided to skip the refactoring of `AsyncCalendarEventNotifier`.

File: app/lib/features/settings/providers/notifiers/devices_notifier.dart [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    final client = ref.watch(alwaysClientProvider);
+    final manager = client.sessionManager();
+
+    _newListener = client.deviceNewEventRx();
+    _newPoller = _newListener?.listen(
+      (evt) async {
+        final sessions = (await manager.allSessions()).toList();
+        state = AsyncValue.data(sessions);
+      },
+      onError: (e, stack) {},
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply app/lib/features/settings/providers/notifiers/devices_notifier.dart]: Fixed by https://github.com/acterglobal/a3/pull/1598/commits/6a165ce352300bdfe7854f1bc66a7dea92559289

File: native/acter/api.rsh [hunk 10]
```diff
@@ -3001,7 +2999,7 @@ object VerificationEvent {
     fn cancel_verification_request() -> Future<Result<bool>>;
 
     /// Bob accepts the verification request from Alice with specified methods
-    fn accept_verification_request_with_methods(methods: Vec<string>) -> Future<Result<bool>>;
+    fn accept_verification_request_with_method(method: string) -> Future<Result<bool>>;
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/api.rsh:3016]: I replaced `Vec<String>` with `String` like [this commit](https://github.com/acterglobal/a3/pull/1950/commits/e074577886590a81739bd76a3d406195f9c16cf2).
Will make it into another PR?

File: native/acter/src/api/verification.rs (hunk 2 above)
Comment (bitfriend) [reply native/acter/src/api/verification.rs:1017]: I just wanted to use `delete_device()` not `delete_devices()`, because argument of this api was changed from `Vec<String>` into `String`.
I found `ruma` provides `delete_device` api and used it.

File: native/acter/src/api/device.rs (hunk 3 above)
Comment (bitfriend) [reply native/acter/src/api/device.rs]: Fixed by https://github.com/acterglobal/a3/pull/1598/commits/943d0005f535bc3b169e251f3c41dc83aabd8732

File: native/acter/src/api/device.rs (hunk 4 above)
Comment (bitfriend) [reply native/acter/src/api/device.rs]: Now detection of new device works fine.
`allSessionsProvider` is updated on every detection.

image

Do you want that acter would pop up dialog like this?

image

File: native/acter/src/api/verification.rs (hunk 2 above)
Comment (bitfriend) [reply native/acter/src/api/verification.rs:1017]: Raised [another PR](https://github.com/acterglobal/a3/pull/2017) and moved this there.

File: native/acter/api.rsh (hunk 10 above)
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/acter/api.rsh:3013]: Raised [another PR](https://github.com/acterglobal/a3/pull/2017) and moved this there.
